### PR TITLE
Optimizer that removes redundant DELIM_GET and DELIM_JOIN operators

### PR DESCRIPTION
**Pull request by lnkuiper**
This PR implements an optimizer which I have affectionately called the __Deliminator__, an optimizer that removes redundant operators related to correlated subqueries.

When a query containing a correlated subquery is issued, a dependent join is created. Consider the following query:
```sql
EXPLAIN SELECT i=ANY(SELECT i FROM integers WHERE i=i1.i) FROM integers i1 ORDER BY i;
```
DuckDB has implemented a way of generically flattening these queries, following _Unnesting Arbitrary Queries_ by _Thomas Neumann_ and _Alfons Kemper_, to avoid quadratic complexity. This is done by creating a _duplicate-eliminated join_, or _delim join_, and pushing it down the query plan to decorrelate it.

This process introduces _delim scans_, which are usually joined (or in a cross product) with another part of the query plan. Under specific circumstances, joining with a delim scan does not introduce new information, and this can be removed. If all delim scans belonging to a delim join can be removed, the delim join can be transformed into a comparison join instead.

This is more efficient for two reasons:
1. Joins with delim scans can be removed, so there is less work to be done
2. Delim joins can be regular comparison joins instead, which are easier (and already) parallelised in DuckDB

For the example query this means that the original query plan
```
┌───────────────────────────┐                                                                                                                    
│         PROJECTION        │                                                                                                                    
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                    
│             #0            │                                                                                                                    
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                    
│             3             │                                                                                                                    
│          (0.00s)          │                                                                                                                    
└─────────────┬─────────────┘                                                                                                                                                 
┌─────────────┴─────────────┐                                                                                                                    
│          ORDER_BY         │                                                                                                                    
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                    
│           #1 ASC          │                                                                                                                    
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                    
│             3             │                                                                                                                    
│          (0.00s)          │                                                                                                                    
└─────────────┬─────────────┘                                                                                                                                                 
┌─────────────┴─────────────┐                                                                                                                    
│         PROJECTION        │                                                                                                                    
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                    
│          SUBQUERY         │                                                                                                                    
│             i             │                                                                                                                    
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                    
│             3             │                                                                                                                    
│          (0.00s)          │                                                                                                                    
└─────────────┬─────────────┘                                                                                                                                                 
┌─────────────┴─────────────┐                                                                                                                    
│         DELIM_JOIN        │                                                                                                                    
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                                                                                    
│            MARK           │                                                                                                                    
│            i=i            │                                                                                                                    
│            i=#0           ├──────────────┐──────────────────────────────────────────────────────────────────────────────────────┐              
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │                                                                                      │              
│             3             │              │                                                                                      │              
│          (0.00s)          │              │                                                                                      │              
└─────────────┬─────────────┘              │                                                                                      │                                           
┌─────────────┴─────────────┐┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐
│          SEQ_SCAN         ││         HASH_JOIN         │                                                          │       HASH_GROUP_BY       │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                          │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          integers         ││            MARK           │                                                          │             #0            │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││            i=i            │                                                          │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             i             ││            i=#0           ├──────────────┐                                           │             0             │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │                                           │          (0.00s)          │
│             3             ││             3             │              │                                           │                           │
│          (0.00s)          ││          (0.00s)          │              │                                           │                           │
└───────────────────────────┘└─────────────┬─────────────┘              │                                           └───────────────────────────┘                             
                             ┌─────────────┴─────────────┐┌─────────────┴─────────────┐                                                          
                             │         CHUNK_SCAN        ││         PROJECTION        │                                                          
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                          
                             │             3             ││             i             │                                                          
                             │          (0.00s)          ││             #0            │                                                          
                             │                           ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                          
                             │                           ││             3             │                                                          
                             │                           ││          (0.00s)          │                                                          
                             └───────────────────────────┘└─────────────┬─────────────┘                                                                                       
                                                          ┌─────────────┴─────────────┐                                                          
                                                          │         HASH_JOIN         │                                                          
                                                          │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                          
                                                          │           INNER           │                                                          
                                                          │            i=i            ├──────────────┐                                           
                                                          │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │                                           
                                                          │             3             │              │                                           
                                                          │          (0.00s)          │              │                                           
                                                          └─────────────┬─────────────┘              │                                                                        
                                                          ┌─────────────┴─────────────┐┌─────────────┴─────────────┐                             
                                                          │         DELIM_SCAN        ││          SEQ_SCAN         │                             
                                                          │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
                                                          │             3             ││          integers         │                             
                                                          │          (0.00s)          ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
                                                          │                           ││             i             │                             
                                                          │                           ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
                                                          │                           ││             3             │                             
                                                          │                           ││          (0.00s)          │                             
                                                          └───────────────────────────┘└───────────────────────────┘                                                          
```

Can be simplified to:
```
┌───────────────────────────┐                             
│         PROJECTION        │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│             #0            │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│             3             │                             
│          (0.00s)          │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│          ORDER_BY         │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│           #1 ASC          │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│             3             │                             
│          (0.00s)          │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│         PROJECTION        │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│          SUBQUERY         │                             
│             i             │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│             3             │                             
│          (0.00s)          │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│         HASH_JOIN         │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│            MARK           │                             
│            i=i            │                             
│            i=#0           ├──────────────┐              
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │              
│             3             │              │              
│          (0.00s)          │              │              
└─────────────┬─────────────┘              │                                           
┌─────────────┴─────────────┐┌─────────────┴─────────────┐
│          SEQ_SCAN         ││         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          integers         ││             i             │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││             i             │
│             i             ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││             3             │
│             3             ││          (0.00s)          │
│          (0.00s)          ││                           │
└───────────────────────────┘└─────────────┬─────────────┘                             
                             ┌─────────────┴─────────────┐
                             │          SEQ_SCAN         │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │          integers         │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │             i             │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │             3             │
                             │          (0.00s)          │
                             └───────────────────────────┘                             
```

The original projection can also be removed in this specific case, but in other cases the projection may project `+(i, 3)`, in which case it is still necessary.

This will improve the performance of the FTS extension, and some TPC-H queries.

I am happy to receive any feedback on this PR!

### CODE REVIEW

File: src/optimizer/deliminator.cpp [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+void Deliminator::FindCandidates(unique_ptr<LogicalOperator> *op_ptr,
+                                 vector<unique_ptr<LogicalOperator> *> &candidates) {
+	auto op = op_ptr->get();
+	// search children before adding, so the deepest candidates get added first
+	for (auto &child : op->children) {
+		FindCandidates(&child, candidates);
+	}
+	if ( // Projection/Aggregate
```
Comment (Mytherin) [inline src/optimizer/deliminator.cpp:99]: I would prefer to break up this huge if statement into several blocks using early outs, e.g.

```c
// search for a projection or aggregate
if (op->type != LogicalOperatorType::LOGICAL_PROJECTION &&
    op->type != LogicalOperatorType::LOGICAL_AGGREGATE_AND_GROUP_BY) {
    return;
}
// followed by a join
if (op->children[0]->type != LogicalOperatorType::LOGICAL_COMPARISON_JOIN) {
    return;
}
...

```

This keeps the same logic but is more readable than a ton of nested conditions (in my opinion).

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! It looks great, and very nice results. I have one minor piece of feedback, otherwise I think it can be merged already.

File: src/optimizer/deliminator.cpp (hunk 1 above)
Comment (lnkuiper) [reply src/optimizer/deliminator.cpp]: Agreed. I split it up and it is much more readable now

**Comment by Mytherin** [comment]
Thanks, looks great now!

**Pull request merged**